packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const {findMetaViewport} = require('../HtmlDomHelper');
+const {calculateHost} = require('../RuntimeHostHelper');
+const validatorRules = require('@ampproject/toolbox-validator-rules');
+
+const BIND_SHORT_FORM_PREFIX = 'bind';
+
+// Some AMP component don't bring their own tag, but enable new attributes on other
+// elements. Most are included in the AMP validation rules, but some are not. These
+// need to be defined manually here.
+const manualAttributeToExtensionMapping = new Map([
+  ['mask', 'amp-inputmask'],
+  ['lightbox', 'amp-lightbox-gallery'],
+]);
+
+/**
+ * Extension Auto Importer - this transformer auto imports all missing AMP extensions.
+ *
+ * The importer analyzes the HTML source code and identifies missing AMP extension imports
+ * using multiple strategies:
+ *
+ * - use validation rules to map used AMP tags to required AMP extensions.
+ * - use validation rules to map used AMP attributes to required AMP extensions.
+ * - manually specifiy attribute to extension mappings if this information is not available in the
+ *   validation rules.
+ * - mnullay implement AMP extension detection for a few corner cases.
+ *
+ * This importer also enables a shortcode `bindtext` instead of `data-amp-bind-text` for specifying
+ * AMP bindings when the square bracket notation (`[text]`) is not available. To avoid accidently
+ * rewriting non-AMP attributes, the transformer uses the AMP validation rules to only rename bindable
+ * attributes as specified in the validation rules.
+ */
+class AutoExtensionImporter {
+  constructor(config) {
+    this.enabled = config.autoExtensionImport !== 'false' && config.autoExtensionImport !== false;
+    this.log_ = config.log.tag('AutoExtensionImporter');
+
+    // We use the validation rules to infer extension imports. The rules are downloaded once and for
+    // efficency, we initially extract all needed rules
+    this.extensionSpec = validatorRules.fetch().then((rules) => {
```

Comment:
Split this into a method (`initSpec` or so). I prefer to avoid non-trivial code in the constructor directly.

Also, consider having the constructor accept (fetched) `validatorRules` instead of fetching it yourself. This allows you to avoid using `await` every single time you access `this.extensionSpec` and makes testing easier.